core/blockchain.go

```diff
@@ -2614,19 +2638,26 @@ func (bc *BlockChain) UpdateValidatorUptime(slots shard.SlotList, mask *bls.Mask
 			stats, err := rawdb.ReadValidatorStats(bc.db, addr)
 			if stats == nil {
 				stats = &staking.ValidatorStats{
-					big.NewInt(0), big.NewInt(0), big.NewInt(0), numeric.NewDec(0),
+					big.NewInt(0), numeric.NewDec(0),
 					[]staking.VotePerShard{}, []staking.KeysPerShard{},
 				}
 			}
-			stats.NumBlocksToSign.Add(stats.NumBlocksToSign, big.NewInt(1))
+			one := big.NewInt(1)
+			wrapper, err := bc.ReadValidatorInformation(addr)
+			if err != nil {
+				return err
+			}
+
+			wrapper.Snapshot.Epoch = epoch
+			wrapper.Snapshot.NumBlocksToSign.Add(wrapper.Snapshot.NumBlocksToSign, one)
 
 			enabled, err := mask.IndexEnabled(i)
 			if err != nil {
 				return err
 			}
 
 			if enabled {
-				stats.NumBlocksSigned.Add(stats.NumBlocksSigned, big.NewInt(1))
+				wrapper.Snapshot.NumBlocksSigned.Add(wrapper.Snapshot.NumBlocksSigned, one)
```

Comment:
As we talked about last time. The update of wrapper needs to be within the block proposal because it's change the state. Here it's just changing the copy of wrapper. And here it's not the right place to change state.

Reply:
Gotcha, I thought about this spot would still count because the increment was happening here and state isn’t written yet

Reply:
Everything in WriteBlockWithState happens after the state is modified.